types/react/index.d.ts

```diff
@@ -666,16 +689,39 @@ declare namespace React {
         defaultProps?: Partial<P>;
     }
 
-    function forwardRef<T, P = {}>(Component: RefForwardingComponent<T, P>): ForwardRefExoticComponent<P & ClassAttributes<T>>;
+    function forwardRef<T, P = {}>(Component: RefForwardingComponent<T, P>): ForwardRefExoticComponent<PropsWithoutRef<P> & RefAttributes<T>>;
 
-    type ComponentProps<T extends ComponentType<any>> =
-        T extends ComponentType<infer P> ? P : {};
-    type ComponentPropsWithRef<T extends ComponentType<any>> =
-        T extends ComponentClass<infer P>
-            ? P & ClassAttributes<InstanceType<T>>
-            : T extends SFC<infer P>
-                ? P
+    /** Ensures that the props do not include ref at all */
+    type PropsWithoutRef<P> =
```

Comment:
I don't think a conditional type is even needed here. We can just do:
```ts
type PropsWithoutRef<P> = Pick<P, Exclude<keyof P, 'ref'>>
```
it will preserve `P` even if `ref` wasn't already there to begin with.

Reply:
I'm trying to avoid an issue with union types -- mapping over a union type will intersect it: https://github.com/Microsoft/TypeScript/issues/28339